### PR TITLE
checkout: Expose new place order process via REST

### PR DESCRIPTION
**Pull request by danielpoe**
Add rest apis to be able to place the current cart

see http://editor.swagger.io/?url=https://raw.githubusercontent.com/i-love-flamingo/flamingo-commerce/more-rest-apis/docs/openapi/swagger.json

### CODE REVIEW

**Comment by carstendietrich** [comment]
Looks promising, will have an in detail look later today.

Guess we would need some integration tests for sure.

Also we are lacking rest routes for the following GraphQL mutations:
```graphql
# Gets the last stored place order state and ensures that the state machine proceeds, non blocking
Commerce_Checkout_RefreshPlaceOrder: Commerce_Checkout_PlaceOrderContext!
# Gets the most recent place order state by waiting for the state machine to proceed, therefore blocking
Commerce_Checkout_RefreshPlaceOrderBlocking: Commerce_Checkout_PlaceOrderContext!
```

Without those we can't proceed in the state machine. And we need definitely add them.

**Comment by danielpoe** [comment]
> Looks promising, will have an in detail look later today.
> 
> Guess we would need some integration tests for sure.
> 
> Also we are lacking rest routes for the following GraphQL mutations:
> 
> ```graphql
> # Gets the last stored place order state and ensures that the state machine proceeds, non blocking
> Commerce_Checkout_RefreshPlaceOrder: Commerce_Checkout_PlaceOrderContext!
> # Gets the most recent place order state by waiting for the state machine to proceed, therefore blocking
> Commerce_Checkout_RefreshPlaceOrderBlocking: Commerce_Checkout_PlaceOrderContext!
> ```
> 
> Without those we can't proceed in the state machine. And we need definitely add them.

Good point - I added the API
I also added integration test :-)

File: checkout/module.go [hunk 1]
```diff
@@ -190,3 +191,27 @@ func (m *Module) Depends() []dingo.Module {
...
+func (r *apiRoutes) Routes(registry *web.RouterRegistry) {
+	registry.Route("/api/v1/checkout/placeorder", "checkout.api.placeorder")
+	registry.HandleGet("checkout.api.placeorder", r.apiController.CurrentPlaceOrderContextAction)
+	registry.HandlePut("checkout.api.placeorder", r.apiController.StartPlaceOrderAction)
+	registry.HandleDelete("checkout.api.placeorder", r.apiController.ClearPlaceOrderAction)
+
+	registry.Route("/api/v1/checkout/placeorder/cancel", "checkout.api.placeorder.cancel")
+	registry.HandleGet("checkout.api.placeorder.cancel", r.apiController.CancelPlaceOrderAction)
+
+	registry.Route("/api/v1/checkout/placeorder/refreshplaceorder", "checkout.api.placeorder.refreshplaceorder")
```
Comment (carstendietrich) [inline checkout/module.go:212]: why duplication of placeorder?
```suggestion
	registry.Route("/api/v1/checkout/placeorder/refresh", "checkout.api.placeorder.refresh")
```

File: checkout/module.go [hunk 2]
```diff
@@ -190,3 +191,27 @@ func (m *Module) Depends() []dingo.Module {
...
+	registry.HandlePost("checkout.api.placeorder.refreshplaceorder", r.apiController.RefreshPlaceOrderAction)
+
+	registry.Route("/api/v1/checkout/placeorder/refreshplaceorderblocking", "checkout.api.placeorder.refreshplaceorderblocking")
```
Comment (carstendietrich) [inline checkout/module.go:215]: maybe:
```suggestion
	registry.Route("/api/v1/checkout/placeorder/refresh/blocking", "checkout.api.placeorder.refresh.blocking")
```

File: test/integrationtest/projecttest/tests/restapi/placeorderszenario_test.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
+// +build integration
```
Comment (carstendietrich) [inline test/integrationtest/projecttest/tests/restapi/placeorderszenario_test.go:1]: please rename file to just `placeorder_test.go`

File: test/integrationtest/projecttest/tests/restapi/placeorderszenario_test.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+	// refresh place order
+	response = e.POST("/api/v1/checkout/placeorder/refreshplaceorderblocking").Expect()
+	response.Status(200).JSON().Object().Value("State").String().NotEmpty()
+
+	// get place order
+	response = e.GET("/api/v1/checkout/placeorder").Expect()
+	response.Status(200).JSON().Object().Value("FailedReason").String().Equal("")
+	response.Status(200).JSON().Object().Value("State").String().Equal("Success")
+
```
Comment (carstendietrich) [inline test/integrationtest/projecttest/tests/restapi/placeorderszenario_test.go:43]: I guess we should at least call each route in this test. Currenty GET /placeorder/ and POST /placeorder/refreshing is missing.

**Comment by carstendietrich** [comment]
> Looks good so far, fixed some typos my self and added missing Swagger summaries.
> Route naming and test coverage could be improved before merging this.

Besides that module readme needs to be updated, previously new process was GraphQL exclusive. CHANGELOG.md in project root should also contain this new additions

File: checkout/interfaces/controller/apicontroller.go [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+	startPlaceOrderResult struct {
+		UUID string
+	}
+
+	// placeOrderContext infos
+	placeOrderContext struct {
+		Cart         *cart.Cart
+		OrderInfos   *placedOrderInfos
+		State        string
+		StateData    interface{}
```
Comment (carstendietrich) [inline checkout/interfaces/controller/apicontroller.go:41]: maybe using type StateData here as well, is also type interface{} but would match the core place order context

File: checkout/interfaces/controller/apicontroller.go [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+// ClearPlaceOrderAction clears the last place order if in final state
+// @Summary Clears the last placed order if in final state
+// @Tags v1 Checkout ajax API
+// @Produce json
+// @Success 200 {boolean} boolean
+// @Failure 500 {object} errorResponse
+// @Router /api/v1/checkout/placeorder [delete]
+func (c *APIController) ClearPlaceOrderAction(ctx context.Context, r *web.Request) web.Result {
+	err := c.placeorderHandler.ClearPlaceOrder(ctx)
+	if err != nil {
```
Comment (carstendietrich) [inline checkout/interfaces/controller/apicontroller.go:160]: should we use another status code whenever there is no place order process running? in this case we currently return `{"Code":"500","Message":"ErrNoPlaceOrderProcess"}`

**Pull request merged**